src/tmio.c

```diff
@@ -408,6 +408,50 @@ TMIO_EHANDSHAKE Error while writing protocol
   return stream->type;
 }
 
+/* Internal helper function to read the protocol string check for match */
+static int tmio_read_protocol(tmio_stream *stream)
+{
+  buftcpfile *fp = (buftcpfile *) stream->f;
+  // Read protocol
+  int protocol_tag;
+  char protocol[TMIO_PROTOCOL_SIZE] = {0};
+  if (buftcpread(&protocol_tag, sizeof(protocol_tag), fp) !=
+          sizeof(protocol_tag) ||
+      protocol_tag != TMIO_PROTOCOL_TAG ||
+      buftcpread(protocol, TMIO_PROTOCOL_SIZE, fp) != TMIO_PROTOCOL_SIZE) {
+    stream->status = TMIO_EHANDSHAKE;
+    if (stream->debug)
+      fprintf(stderr, "tmio_read_protocol: protocol handshake failed,"
+              "received protocol tag %d with protocol string %s\n",
+              protocol_tag, protocol);
+
+    tmio_close(stream);
+    return -1;
+  }
+  stream->bytesread += sizeof(protocol_tag) + TMIO_PROTOCOL_SIZE;
+
+  // Sanitise input
+  protocol[TMIO_PROTOCOL_SIZE - 1] = 0;
+  const int protocol_len = strlen(stream->protocol);
+
+  // Compare up to strlen(requested protocol) bytes
+  if (strncmp(protocol, stream->protocol, protocol_len) != 0) {
+    stream->status = TMIO_EPROTO;
+    if (stream->debug)
+      fprintf(stderr, "tmio_read_protocol: peer has wrong protocol %s,"
+              "expected %s\n", protocol, stream->protocol);
+    return -1;
+  }
+
+  if (stream->debug > 1)
+    fprintf(stderr, "tmio_read_protocol: protocol handshake successful,"
+            "received protocol tag %d with protocol string %s\n",
+            protocol_tag, protocol);
+
+  // Copy hidden messages in protocol string ;)
+  strncpy(stream->protocol, protocol, TMIO_PROTOCOL_SIZE);
```

Comment:
This breaks the original idea that we only match the protocol prefix, e.g. if we request "FlashCamV1" we should be able to read all "FlashCamV1.x" files.

Reply:
Isn't this the same as in the previous version?
Line 438 here compares until protocol_len of `stream-protocol` and then copies the full string? (see line 495 in the old code). I just removed the ternary operator because the sanitisation of the input already requires that `strlen(stream->protocol) < TMIO_PROTOCOL_SIZE`?
I thought I only added a snarky comment, I could change that though :)
Maybe I overlook something..